### PR TITLE
Add support for tuning a receive buffer of the UDP socket

### PR DESCRIPTION
**Pull request by hasso**
In case of larger collectd traffic there is a need to increase a receive buffer of the UDP socket to avoid drops. Increasing defaults in OS is not often what you really want to do, as it changes default for all sockets in the system.

With go-collectd you can create your own socket, apply all needed options and pass it to the Server (https://github.com/collectd/go-collectd/issues/22).

Probably there should be defer to close the socket as well somewhere, but I'm not really sure how it should be used with goroutines.

### CODE REVIEW

**Comment by hasso** [comment]
No comments?

**Comment by juliusv** [comment]
@hasso Sorry and thanks for the ping. This fells through the cracks.

I'm wondering, the default connection setup in your new vendoring (when the user doesn't specify their own `Conn`) has a few things which are omitted in your own setup, but which seem sensible: https://github.com/hasso/collectd_exporter/blob/b177f50aa861d5db6e919bc667d73240d9674a08/vendor/collectd.org/network/server.go#L43-L70

- the setting of the `addr` if it's empty
- setting the multicast interface name (though not sure of the significance)

Maybe add those to your custom connection setup too?

File: main.go [hunk 1]
```diff
@@ -248,6 +250,25 @@ func startCollectdServer(ctx context.Context, w api.Writer) {
 		log.Fatalf("Unknown security level %q. Must be one of \"None\", \"Sign\" and \"Encrypt\".", *collectdSecurity)
 	}
 
+	laddr, err := net.ResolveUDPAddr("udp", *collectdAddress)
+	if err != nil {
+		log.Fatal(err)
```
Comment (juliusv) [inline main.go]: These `log.Fatal`s should have some wrapper error message before the raw returned error, like the `log.Fatal` above in line 250. In this case here, something like:

```go
log.Fatalf("Failed to resolve binary protocol listening UDP address %q: %v", *collectdAddress, err)
```

File: main.go [hunk 2]
```diff
@@ -248,6 +250,25 @@ func startCollectdServer(ctx context.Context, w api.Writer) {
...
+	}
+
+	if laddr.IP != nil && laddr.IP.IsMulticast() {
+		srv.Conn, err = net.ListenMulticastUDP("udp", nil, laddr)
+	} else {
+		srv.Conn, err = net.ListenUDP("udp", laddr)
+	}
+	if err != nil {
+		log.Fatal(err)
```
Comment (juliusv) [inline main.go]: Same here and below.

**Comment by juliusv** [comment]
Never closing the connection is probably ok in this case, as it would only get closed when the program exits anyways.

**Comment by hasso** [comment]
I made log messages more informative now. The code already handles the case when address is empty – in collectd_exporter context this means that binary protocol server isn't started at all. Which is completely valid use case of course.

I'm not sure about multicast – I don't use it and I'm not even sure that it should be there. Yes, I know that collectd supports that, but I consider such use a problematic can of worms in practice. Adding support for specifying interface is easy, but it adds another command line option which wouldn't be used in reality, probably. If you say it's OK, I can add the code of course.

**Comment by juliusv** [comment]
Good point about the empty address!

About multicast: no real idea either, but yeah, after reading the code once again I tend to agree.

Thanks! :+1:

**Pull request merged**